sierra_adapter/sierra_indexer/src/main/scala/weco/catalogue/sierra_indexer/services/Splitter.scala

```diff
@@ -0,0 +1,156 @@
+// that can be sent to Elasticsearch.
+class Splitter(indexPrefix: String) {
+  import JsonOps._
+
+  private val varFieldIndex = Index(s"${indexPrefix}_varfields")
+  private val fixedFieldsIndex = Index(s"${indexPrefix}_fixedfields")
+
+  implicit val encoder: Encoder[SierraRecordTypes.Value] =
+    (value: SierraRecordTypes.Value) => Json.fromString(value.toString)
+
+  implicit val encoderNumber: Encoder[TypedSierraRecordNumber] =
+    (number: TypedSierraRecordNumber) =>
+      Json.fromString(number.withoutCheckDigit)
+
+  implicit val encoderParent: Encoder[Parent] = deriveConfiguredEncoder
+
+  private case class IndexedVarField(parent: Parent,
+                                     position: Int,
+                                     varField: Json)
+  private case class IndexedFixedField(parent: Parent,
+                                       code: String,
+                                       fixedField: Json)
+
+  def split(t: SierraTransformable)
+    : Either[Seq[(Parent, ParsingFailure)],
+             (Seq[IndexRequest], Seq[DeleteByQueryRequest])] = {
+    for {
+      apiData <- getSierraApiData(t)
+
+      mainRecords = List(
+        apiData.map {
+          case (parent, json) =>
+            IndexRequest(
+              index = Index(s"${indexPrefix}_${parent.recordType}"),
+              id = Some(parent.id.withoutCheckDigit),
+              source = Some(json.remainder.noSpaces)
+            )
+        }
+      ).flatten
+
+      varFields = apiData.flatMap {
+        case (parent, json) =>
+          json.varFields.zipWithIndex
+            .map {
+              case (varField, position) =>
+                IndexRequest(
+                  index = varFieldIndex,
+                  id = Some(s"${parent.id}-$position"),
+                  source = Some(
+                    IndexedVarField(parent, position, varField).asJson.noSpaces
+                  )
+                )
+            }
+      }
+
+      varFieldDeletions = apiData.map {
+        case (parent, json) =>
+          DeleteByQueryRequest(
+            indexes = Indexes(varFieldIndex.name),
+            query = must(
+              termQuery("parent.id.keyword", parent.id),
+              termQuery(
+                "parent.recordType.keyword",
+                parent.recordType.toString),
+              rangeQuery("position").gte(json.varFields.length)
+            )
+          )
+      }
+
+      fixedFields = apiData.flatMap {
+        case (parent, json) =>
+          json.fixedFields
+            .map {
+              case (code, fixedField) =>
+                IndexRequest(
+                  index = fixedFieldsIndex,
+                  id = Some(s"${parent.id}-$code"),
+                  source = Some(
+                    IndexedFixedField(parent, code, fixedField).asJson.noSpaces
+                  )
+                )
+            }
+      }
+
+      fixedFieldDeletions = apiData.map {
```

Comment:
Can the functions inside these `apiData.map/flatMap`s be extracted into methods so we can write something like:
```scala
for {
  apiData <- getSierraApiData(t)
  mainRecords = apiData.flatMap(IndexRequest.mainRecords)
  varFields = apiData.flatMap(IndexRequest.varFields)
  varFieldDeletions = apiData.map(IndexRequest.varFieldDeletions)
  fixedFields = apiData.flatMap(IndexRequest.fixedFields)
  fixedFieldDeletions = apiData.map(IndexRequest.fixedFieldDeletions)
} yield (
  mainRecords ++ varFields ++ fixedFields,
  varFieldDeletions ++ fixedFieldDeletions
)
```
They're a bit hard to follow and it would clarify that they are not interdependent (unless they are and I've missed that, in which case they could do with being more explicit anyway)

Reply:
Good shout, changed. Hopefully it’s a bit clearer now (and easier to test, too!).